src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1519,14 +1602,66 @@ private BoundExpression GetDefaultParameterSpecialNoConversion(SyntaxNode syntax
                     BoundCollectionExpression converted;
                     if (!conversion.Exists)
                     {
-                        // PROTOTYPE(ParamsCollections): Add test if this code path is reachable
+                        Debug.Assert(false); // Add test if this code path is reachable
                         GenerateImplicitConversionErrorForCollectionExpression(unconvertedCollection, collectionType, diagnostics);
                         converted = BindCollectionExpressionForErrorRecovery(unconvertedCollection, collectionType, inConversion: true, diagnostics);
                     }
                     else
                     {
                         Debug.Assert(conversion.IsCollectionExpression);
-                        converted = ConvertCollectionExpression(unconvertedCollection, collectionType, conversion, diagnostics);
+
+                        bool infiniteRecursion = false;
+                        if (conversion.GetCollectionExpressionTypeKind(out _, out MethodSymbol? constructor, out bool isExpanded) == CollectionExpressionTypeKind.ImplementsIEnumerable &&
+                            isExpanded)
+                        {
+                            Debug.Assert(constructor is not null);
+
+                            // Check for infinite recursion through the constructor
+                            var constructorSet = PooledHashSet<MethodSymbol>.GetInstance();
+                            constructorSet.Add(constructor.OriginalDefinition);
+
+                            BoundUnconvertedCollectionExpression? emptyCollection = null;
+
+                            while (true)
+                            {
+                                var paramsType = constructor.Parameters[^1].Type;
+                                if (!paramsType.IsSZArray())
+                                {
+                                    var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+                                    emptyCollection ??= new BoundUnconvertedCollectionExpression(node, ImmutableArray<BoundNode>.CastUp(ImmutableArray<BoundExpression>.Empty)) { WasCompilerGenerated = true, IsParamsCollection = true };
```

Comment:
>ImmutableArray<BoundNode>.CastUp(ImmutableArray<BoundExpression>.Empty) [](http://example.com/codeflow?start=103&length=71)

`ImmutableArray<BoundNode>.Empty`?

Reply:
No, we need this array of ```BoundExpression``` underneath, so that we could do a cast down without allocation/copy. Perhaps for an empty array, there isn't much to allocate and copy, but I prefer to be consistent